snuba_sdk/dsl/dsl.py

```diff
@@ -0,0 +1,296 @@
+)
+MRI_NAME_REGEX = r"([a-z_]+(?:\.[a-z_]+)*)"
+UNIT_REGEX = r"([\w.]*)"
+GRAMMAR = Grammar(
+    rf"""
+expression = term (_ expr_op _ term)*
+expr_op = "+" / "-"
+
+term = coefficient (_ term_op _ coefficient)*
+term_op = "*" / "/"
+coefficient = number / filter
+
+number = ~r"[0-9]+" ("." ~r"[0-9]+")?
+filter = target ("{{" _ condition (_ "," _ condition)* _ "}}")? (group_by)?
+
+condition = (variable / tag_key) _ condition_op _ tag_value
+condition_op = "=" / "!=" / "~" / "!~" / "IN" / "NOT IN"
+tag_key = ~"[a-zA-Z0-9_]+"
+tag_value = quoted_string / quoted_string_tuple / variable
+
+quoted_string = ~r'"([^"\\]*(?:\\.[^"\\]*)*)"'
+quoted_string_tuple = "(" _ quoted_string (_ "," _ quoted_string)* _ ")"
+
+target = variable / nested_expression / function / metric
+variable = "$" ~"[a-zA-Z0-9_]+"
+nested_expression = "(" _ expression _ ")"
+
+function = aggregate (group_by)?
+aggregate = aggregate_name ("(" _ expression (_ "," _ expression)* _ ")")
+aggregate_name = ~"[a-zA-Z0-9_]+"
+
+group_by = _ "by" _ (group_by_name / group_by_name_tuple)
+group_by_name = ~"[a-zA-Z0-9_]+"
+group_by_name_tuple = "(" _ group_by_name (_ "," _ group_by_name)* _ ")"
+
+metric = quoted_mri / unquoted_mri / quoted_public_name / unquoted_public_name
+quoted_mri = ~r'`{ENTITY_TYPE_REGEX}:{NAMESPACE_REGEX}/{MRI_NAME_REGEX}@{UNIT_REGEX}`'
+unquoted_mri = ~r'{ENTITY_TYPE_REGEX}:{NAMESPACE_REGEX}/{MRI_NAME_REGEX}@{UNIT_REGEX}'
+quoted_public_name = ~r'`([a-z_]+(?:\.[a-z_]+)*)`'
+unquoted_public_name = ~r'([a-z_]+(?:\.[a-z_]+)*)'
+
+_ = ~r"\s*"
+"""
+)
+
+EXPRESSION_OPERATORS: Mapping[str, str] = {
+    "+": ArithmeticFunction.PLUS.value,
+    "-": ArithmeticFunction.MINUS.value,
+}
+
+TERM_OPERATORS: Mapping[str, str] = {
+    "*": ArithmeticFunction.MULTIPLY.value,
+    "/": ArithmeticFunction.DIVIDE.value,
+}
+
+
+def parse_mql(mql: str) -> MetricsQuery:
+    """
+    Parse a MQL string into a MetricsQuery object.
+    """
+    try:
+        tree = GRAMMAR.parse(mql.strip())
+    except ParseError as e:
+        raise InvalidQueryError("Invalid metrics syntax") from e
+    result = MQLlVisitor().visit(tree)
+    assert isinstance(result, MetricsQuery)
+    return result
+
+
+class MQLlVisitor(NodeVisitor):
+    def visit(self, node: Node) -> Any:
+        """Walk a parse tree, transforming it into a MetricsQuery object.
+
+        Recursively descend a parse tree, dispatching to the method named after
+        the rule in the :class:`~parsimonious.grammar.Grammar` that produced
+        each node. If, for example, a rule was... ::
+
+            bold = '<b>'
+
+        ...the ``visit_bold()`` method would be called.
+        """
+        method = getattr(self, "visit_" + node.expr_name, self.generic_visit)
+        try:
+            result = method(node, [self.visit(n) for n in node])
+            return result
+        except Exception:
+            raise Exception
+
+    def collapse_into_timeseries(self, metric_query: MetricsQuery) -> Timeseries:
```

Comment:
Why do you do this? Why not leave the filters/groupby at the `MetricsQuery` level? We'll end up with duplicated conditions. And when the query is sent to Snuba later, it will still have to push all the top level fields down to the timeseries again, in case any new timeseries were added/changed.

Reply:
Ah I see, yes that does make sense.

I think this might be solved by not resolving things into a MetricsQuery until the entire DSL has been parsed. That way each component of the string is always guaranteed to return a Timeseries. Only after the entire string has been parsed into a Function or a Timeseries does it get wrapped up into a MetricsQuery.

Reply:
As discussed offline, we may need to rework how the MetricsQuery, Timeseries, and Function classes look like. Currently, the MetricsQuery contains information about the top level filters and groupbys which requires it to resolve before the entire DSL is parsed given the way the nodes are traversed. We should move these top-level attributes elsewhere such that parsing the DSL resolves into a singular Timeseries/Equation.

Reply:
Added in https://github.com/getsentry/snuba-sdk/pull/127.